Fedora Package Build Fails SELinux Related Unit Tests

I am a Fedora package maintainer and I am working on getting umoci into the distro.  One core tenet of packaging golang projects in Fedora is that we MUST[1] run unit tests.  Package builds of umoci fail with the following output:
```
--- FAIL: TestUnpackGenerateRoundTrip_ComplexXattr_OverlayfsRootfs (0.03s)
    --- FAIL: TestUnpackGenerateRoundTrip_ComplexXattr_OverlayfsRootfs/TarEntries=user.overlay. (0.02s)
        --- FAIL: TestUnpackGenerateRoundTrip_ComplexXattr_OverlayfsRootfs/TarEntries=user.overlay./DirRootfs (0.01s)
            xattr_linux_test.go:216: 
                	Error Trace:	/builddir/build/BUILD/umoci-0.5.0-build/umoci-0.5.0/src/github.com/opencontainers/umoci/oci/layer/xattr_linux_test.go:216
                	Error:      	Not equal: 
                	            	expected: map[string]string{"user.dummy.xattr":"foobar", "user.overlay.opaque":"x"}
                	            	actual  : map[string]string{"security.selinux":"system_u:object_r:mock_var_lib_t:s0\x00", "user.dummy.xattr":"foobar", "user.overlay.opaque":"x"}
                	            	
                	            	Diff:
                	            	--- Expected
                	            	+++ Actual
                	            	@@ -1,2 +1,3 @@
                	            	-(map[string]string) (len=2) {
                	            	+(map[string]string) (len=3) {
                	            	+ (string) (len=16) "security.selinux": (string) (len=36) "system_u:object_r:mock_var_lib_t:s0\x00",
                	            	  (string) (len=16) "user.dummy.xattr": (string) (len=6) "foobar",
                	Test:       	TestUnpackGenerateRoundTrip_ComplexXattr_OverlayfsRootfs/TarEntries=user.overlay./DirRootfs
                	Messages:   	UnpackEntry("."): expected to see map[string]string{"user.dummy.xattr":"foobar", "user.overlay.opaque":"x"} not be remapped
            xattr_linux_test.go:216: 
                	Error Trace:	/builddir/build/BUILD/umoci-0.5.0-build/umoci-0.5.0/src/github.com/opencontainers/umoci/oci/layer/xattr_linux_test.go:216
                	Error:      	Not equal: 
                	            	expected: map[string]string{"user.overlay.fakexattr":"fakexattr"}
                	            	actual  : map[string]string{"security.selinux":"system_u:object_r:mock_var_lib_t:s0\x00", "user.overlay.fakexattr":"fakexattr"}
                	            	
                	            	Diff:
                	            	--- Expected
                	            	+++ Actual
                	            	@@ -1,2 +1,3 @@
                	            	-(map[string]string) (len=1) {
                	            	+(map[string]string) (len=2) {
                	            	+ (string) (len=16) "security.selinux": (string) (len=36) "system_u:object_r:mock_var_lib_t:s0\x00",
                	            	  (string) (len=22) "user.overlay.fakexattr": (string) (len=9) "fakexattr"
                	Test:       	TestUnpackGenerateRoundTrip_ComplexXattr_OverlayfsRootfs/TarEntries=user.overlay./DirRootfs
                	Messages:   	UnpackEntry("foo/"): expected to see map[string]string{"user.overlay.fakexattr":"fakexattr"} not be remapped
            xattr_linux_test.go:216: 
                	Error Trace:	/builddir/build/BUILD/umoci-0.5.0-build/umoci-0.5.0/src/github.com/opencontainers/umoci/oci/layer/xattr_linux_test.go:216
                	Error:      	Not equal: 
                	            	expected: map[string]string{"user.overlay.whiteout":"foo"}
                	            	actual  : map[string]string{"security.selinux":"system_u:object_r:mock_var_lib_t:s0\x00", "user.overlay.whiteout":"foo"}
                	            	
                	            	Diff:
                	            	--- Expected
                	            	+++ Actual
                	            	@@ -1,2 +1,3 @@
                	            	-(map[string]string) (len=1) {
                	            	+(map[string]string) (len=2) {
                	            	+ (string) (len=16) "security.selinux": (string) (len=36) "system_u:object_r:mock_var_lib_t:s0\x00",
                	            	  (string) (len=21) "user.overlay.whiteout": (string) (len=3) "foo"
                	Test:       	TestUnpackGenerateRoundTrip_ComplexXattr_OverlayfsRootfs/TarEntries=user.overlay./DirRootfs
                	Messages:   	UnpackEntry("foo/bar"): expected to see map[string]string{"user.overlay.whiteout":"foo"} not be remapped
            xattr_linux_test.go:216: 
                	Error Trace:	/builddir/build/BUILD/umoci-0.5.0-build/umoci-0.5.0/src/github.com/opencontainers/umoci/oci/layer/xattr_linux_test.go:216
                	Error:      	Not equal: 
                	            	expected: map[string]string{"user.overlay.opaque":"y"}
                	            	actual  : map[string]string{"security.selinux":"system_u:object_r:mock_var_lib_t:s0\x00", "user.overlay.opaque":"y"}
                	            	
                	            	Diff:
                	            	--- Expected
                	            	+++ Actual
                	            	@@ -1,2 +1,3 @@
                	            	-(map[string]string) (len=1) {
                	            	+(map[string]string) (len=2) {
                	            	+ (string) (len=16) "security.selinux": (string) (len=36) "system_u:object_r:mock_var_lib_t:s0\x00",
                	            	  (string) (len=19) "user.overlay.opaque": (string) (len=1) "y"

```
The above output is only a snippet.  There appear to be approximately 24 failures related to `xattr_linux_test.go`.

Please note:
When I run the package build on my local system _with selinux disabled_, I get no such errors.  When _selinux is enabled_, these tests fail.  The Fedora build environment runs with SELinux enabled, and to my knowledge, is unchangeable.







[1] - https://docs.fedoraproject.org/en-US/packaging-guidelines/Golang/#_testing

@cyphar, here is the error output.  This looks to be only related to SELinux xattrs:
```
Executing(%check): /bin/sh -e /var/tmp/rpm-tmp.jnTCm2
+ umask 022
+ cd /builddir/build/BUILD/umoci-0.5.0-build
+ cd umoci-0.5.0
+ export GOPATH=/builddir/build/BUILD/umoci-0.5.0-build/umoci-0.5.0
+ GOPATH=/builddir/build/BUILD/umoci-0.5.0-build/umoci-0.5.0
+ cd src/github.com/opencontainers/umoci
+ GO111MODULE=off
+ go test -buildmode pie -compiler gc -ldflags '  -extldflags '\''-Wl,-z,relro -Wl,--as-needed  -Wl,-z,pack-relative-relocs -Wl,-z,now -specs=/usr/lib/rpm/redhat/redhat-hardened-ld -specs=/usr/lib/rpm/redhat/redhat-hardened-ld-errors -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1  -Wl,--build-id=sha1 -specs=/usr/lib/rpm/redhat/redhat-package-notes   '\''' ./...
ok  	github.com/opencontainers/umoci	0.011s
ok  	github.com/opencontainers/umoci/cmd/umoci	0.010s
?   	github.com/opencontainers/umoci/internal	[no test files]
ok  	github.com/opencontainers/umoci/mutate	1.433s
?   	github.com/opencontainers/umoci/oci/cas	[no test files]
ok  	github.com/opencontainers/umoci/oci/cas/dir	0.017s
ok  	github.com/opencontainers/umoci/oci/casext	3.543s
ok  	github.com/opencontainers/umoci/oci/casext/blobcompress	0.020s
?   	github.com/opencontainers/umoci/oci/casext/mediatype	[no test files]
?   	github.com/opencontainers/umoci/oci/config/convert	[no test files]
ok  	github.com/opencontainers/umoci/oci/config/generate	0.009s
2025/08/06 17:55:14  warn generate layer: could not add file "some/parents/filechanged": add file lstat: lstat /tmp/TestGenerateMissingFileError1579191314/001/some/parents/filechanged: no such file or directory
2025/08/06 17:55:14  warn could not generate layer: generate layer file: add file lstat: lstat /tmp/TestGenerateMissingFileError1579191314/001/some/parents/filechanged: no such file or directory
2025/08/06 17:55:14  warn generate layer: could not add file "some/parents": add file lstat: lstat /tmp/TestGenerateWrongRootError1834968909/001/some/some/parents: no such file or directory
2025/08/06 17:55:14  warn could not generate layer: generate layer file: add file lstat: lstat /tmp/TestGenerateWrongRootError1834968909/001/some/some/parents: no such file or directory
2025/08/06 17:55:14  info unpack rootfs: /tmp/TestUnpackManifestCustomLayer730166024/002/rootfs
2025/08/06 17:55:14  info unpack layer: sha256:159fc03d42b1eb2588cdffaabe13b9a4b25879d64db0ea473c6c411adcc5740a
2025/08/06 17:55:14  info unpack layer: sha256:29788ced4c99508cba36d0a32e9029789641b2996dbead688cc114fc7203cd9c
2025/08/06 17:55:14  info unpack rootfs: /tmp/TestUnpackStartFromDescriptor407066891/002/rootfs
2025/08/06 17:55:14  info unpack layer: sha256:29788ced4c99508cba36d0a32e9029789641b2996dbead688cc114fc7203cd9c
2025/08/06 17:55:14  info unpack rootfs: /tmp/TestUnpackUnimplementedOverlayfs1799452549/002/rootfs
2025/08/06 17:55:14  info image contains layers using the deprecated 'non-distributable' media type "application/vnd.oci.image.layer.nondistributable.v1.tar"
2025/08/06 17:55:14  info image contains layers using the deprecated 'non-distributable' media type "application/vnd.oci.image.layer.nondistributable.v1.tar"
2025/08/06 17:55:14  info image contains layers using the deprecated 'non-distributable' media type "application/vnd.oci.image.layer.nondistributable.v1.tar"
2025/08/06 17:55:14  info image contains layers using the deprecated 'non-distributable' media type "application/vnd.oci.image.layer.nondistributable.v1.tar"
--- FAIL: TestUnpackGenerateRoundTrip_ComplexXattr_OverlayfsRootfs (0.03s)
    --- FAIL: TestUnpackGenerateRoundTrip_ComplexXattr_OverlayfsRootfs/TarEntries=user.overlay. (0.02s)
        --- FAIL: TestUnpackGenerateRoundTrip_ComplexXattr_OverlayfsRootfs/TarEntries=user.overlay./DirRootfs (0.01s)
            xattr_linux_test.go:216: 
                	Error Trace:	/builddir/build/BUILD/umoci-0.5.0-build/umoci-0.5.0/src/github.com/opencontainers/umoci/oci/layer/xattr_linux_test.go:216
                	Error:      	Not equal: 
                	            	expected: map[string]string{"user.dummy.xattr":"foobar", "user.overlay.opaque":"x"}
                	            	actual  : map[string]string{"security.selinux":"system_u:object_r:mock_var_lib_t:s0\x00", "user.dummy.xattr":"foobar", "user.overlay.opaque":"x"}
                	            	
                	            	Diff:
                	            	--- Expected
                	            	+++ Actual
                	            	@@ -1,2 +1,3 @@
                	            	-(map[string]string) (len=2) {
                	            	+(map[string]string) (len=3) {
                	            	+ (string) (len=16) "security.selinux": (string) (len=36) "system_u:object_r:mock_var_lib_t:s0\x00",
                	            	  (string) (len=16) "user.dummy.xattr": (string) (len=6) "foobar",
                	Test:       	TestUnpackGenerateRoundTrip_ComplexXattr_OverlayfsRootfs/TarEntries=user.overlay./DirRootfs
                	Messages:   	UnpackEntry("."): expected to see map[string]string{"user.dummy.xattr":"foobar", "user.overlay.opaque":"x"} not be remapped
            xattr_linux_test.go:216: 
                	Error Trace:	/builddir/build/BUILD/umoci-0.5.0-build/umoci-0.5.0/src/github.com/opencontainers/umoci/oci/layer/xattr_linux_test.go:216
                	Error:      	Not equal: 
                	            	expected: map[string]string{"user.overlay.fakexattr":"fakexattr"}
                	            	actual  : map[string]string{"security.selinux":"system_u:object_r:mock_var_lib_t:s0\x00", "user.overlay.fakexattr":"fakexattr"}
                	            	
                	            	Diff:
                	            	--- Expected
                	            	+++ Actual
                	            	@@ -1,2 +1,3 @@
                	            	-(map[string]string) (len=1) {
                	            	+(map[string]string) (len=2) {
                	            	+ (string) (len=16) "security.selinux": (string) (len=36) "system_u:object_r:mock_var_lib_t:s0\x00",
                	            	  (string) (len=22) "user.overlay.fakexattr": (string) (len=9) "fakexattr"
                	Test:       	TestUnpackGenerateRoundTrip_ComplexXattr_OverlayfsRootfs/TarEntries=user.overlay./DirRootfs
                	Messages:   	UnpackEntry("foo/"): expected to see map[string]string{"user.overlay.fakexattr":"fakexattr"} not be remapped
            xattr_linux_test.go:216: 
                	Error Trace:	/builddir/build/BUILD/umoci-0.5.0-build/umoci-0.5.0/src/github.com/opencontainers/umoci/oci/layer/xattr_linux_test.go:216
                	Error:      	Not equal: 
                	            	expected: map[string]string{"user.overlay.whiteout":"foo"}
                	            	actual  : map[string]string{"security.selinux":"system_u:object_r:mock_var_lib_t:s0\x00", "user.overlay.whiteout":"foo"}
                	            	
                	            	Diff:
                	            	--- Expected
                	            	+++ Actual
                	            	@@ -1,2 +1,3 @@
                	            	-(map[string]string) (len=1) {
                	            	+(map[string]string) (len=2) {
                	            	+ (string) (len=16) "security.selinux": (string) (len=36) "system_u:object_r:mock_var_lib_t:s0\x00",
                	            	  (string) (len=21) "user.overlay.whiteout": (string) (len=3) "foo"
                	Test:       	TestUnpackGenerateRoundTrip_ComplexXattr_OverlayfsRootfs/TarEntries=user.overlay./DirRootfs
                	Messages:   	UnpackEntry("foo/bar"): expected to see map[string]string{"user.overlay.whiteout":"foo"} not be remapped
            xattr_linux_test.go:216: 
                	Error Trace:	/builddir/build/BUILD/umoci-0.5.0-build/umoci-0.5.0/src/github.com/opencontainers/umoci/oci/layer/xattr_linux_test.go:216
                	Error:      	Not equal: 
                	            	expected: map[string]string{"user.overlay.opaque":"y"}
                	            	actual  : map[string]string{"security.selinux":"system_u:object_r:mock_var_lib_t:s0\x00", "user.overlay.opaque":"y"}
                	            	
                	            	Diff:
                	            	--- Expected
                	            	+++ Actual
                	            	@@ -1,2 +1,3 @@
                	            	-(map[string]string) (len=1) {
                	            	+(map[string]string) (len=2) {
                	            	+ (string) (len=16) "security.selinux": (string) (len=36) "system_u:object_r:mock_var_lib_t:s0\x00",
                	            	  (string) (len=19) "user.overlay.opaque": (string) (len=1) "y"
                	Test:       	TestUnpackGenerateRoundTrip_ComplexXattr_OverlayfsRootfs/TarEntries=user.overlay./DirRootfs
                	Messages:   	UnpackEntry("foo/baz/"): expected to see map[string]string{"user.overlay.opaque":"y"} not be remapped
            xattr_linux_test.go:216: 
                	Error Trace:	/builddir/build/BUILD/umoci-0.5.0-build/umoci-0.5.0/src/github.com/opencontainers/umoci/oci/layer/xattr_linux_test.go:216
                	Error:      	Not equal: 
                	            	expected: map[string]string{"user.overlay.overlay.opaque":"x", "user.overlay.overlay.overlay.overlay.overlay.overlay.dummy":"dummy xattr", "user.overlay.overlay.overlay.whiteout":"foobar"}
                	            	actual  : map[string]string{"security.selinux":"system_u:object_r:mock_var_lib_t:s0\x00", "user.overlay.overlay.opaque":"x", "user.overlay.overlay.overlay.overlay.overlay.overlay.dummy":"dummy xattr", "user.overlay.overlay.overlay.whiteout":"foobar"}
                	            	
                	            	Diff:
                	            	--- Expected
                	            	+++ Actual
                	            	@@ -1,2 +1,3 @@
                	            	-(map[string]string) (len=3) {
                	            	+(map[string]string) (len=4) {
                	            	+ (string) (len=16) "security.selinux": (string) (len=36) "system_u:object_r:mock_var_lib_t:s0\x00",
                	            	  (string) (len=27) "user.overlay.overlay.opaque": (string) (len=1) "x",
                	Test:       	TestUnpackGenerateRoundTrip_ComplexXattr_OverlayfsRootfs/TarEntries=user.overlay./DirRootfs
                	Messages:   	UnpackEntry("foo/extra-nesting/"): expected to see map[string]string{"user.overlay.overlay.opaque":"x", "user.overlay.overlay.overlay.overlay.overlay.overlay.dummy":"dummy xattr", "user.overlay.overlay.overlay.whiteout":"foobar"} not be remapped
            xattr_linux_test.go:216: 
                	Error Trace:	/builddir/build/BUILD/umoci-0.5.0-build/umoci-0.5.0/src/github.com/opencontainers/umoci/oci/layer/xattr_linux_test.go:216
                	Error:      	Not equal: 
                	            	expected: map[string]string{"user.overlay.overlay.overlay.overlay.overlay.dummy123":"dummy xattr 123"}
                	            	actual  : map[string]string{"security.selinux":"system_u:object_r:mock_var_lib_t:s0\x00", "user.overlay.overlay.overlay.overlay.overlay.dummy123":"dummy xattr 123"}
                	            	
                	            	Diff:
                	            	--- Expected
                	            	+++ Actual
                	            	@@ -1,2 +1,3 @@
                	            	-(map[string]string) (len=1) {
                	            	+(map[string]string) (len=2) {
                	            	+ (string) (len=16) "security.selinux": (string) (len=36) "system_u:object_r:mock_var_lib_t:s0\x00",
                	            	  (string) (len=53) "user.overlay.overlay.overlay.overlay.overlay.dummy123": (string) (len=15) "dummy xattr 123"
                	Test:       	TestUnpackGenerateRoundTrip_ComplexXattr_OverlayfsRootfs/TarEntries=user.overlay./DirRootfs
                	Messages:   	UnpackEntry("foo/extra-nesting/reg"): expected to see map[string]string{"user.overlay.overlay.overlay.overlay.overlay.dummy123":"dummy xattr 123"} not be remapped
        --- FAIL: TestUnpackGenerateRoundTrip_ComplexXattr_OverlayfsRootfs/TarEntries=user.overlay./OverlayfsRootfs-TrustedXattr (0.01s)
            xattr_linux_test.go:216: 
                	Error Trace:	/builddir/build/BUILD/umoci-0.5.0-build/umoci-0.5.0/src/github.com/opencontainers/umoci/oci/layer/xattr_linux_test.go:216
                	Error:      	Not equal: 
                	            	expected: map[string]string{"user.dummy.xattr":"foobar", "user.overlay.opaque":"x"}
                	            	actual  : map[string]string{"security.selinux":"system_u:object_r:mock_var_lib_t:s0\x00", "user.dummy.xattr":"foobar", "user.overlay.opaque":"x"}
                	            	
                	            	Diff:
                	            	--- Expected
                	            	+++ Actual
                	            	@@ -1,2 +1,3 @@
                	            	-(map[string]string) (len=2) {
                	            	+(map[string]string) (len=3) {
                	            	+ (string) (len=16) "security.selinux": (string) (len=36) "system_u:object_r:mock_var_lib_t:s0\x00",
                	            	  (string) (len=16) "user.dummy.xattr": (string) (len=6) "foobar",
                	Test:       	TestUnpackGenerateRoundTrip_ComplexXattr_OverlayfsRootfs/TarEntries=user.overlay./OverlayfsRootfs-TrustedXattr
                	Messages:   	UnpackEntry("."): expected to see map[string]string{"user.dummy.xattr":"foobar", "user.overlay.opaque":"x"} not be remapped
            xattr_linux_test.go:216: 
                	Error Trace:	/builddir/build/BUILD/umoci-0.5.0-build/umoci-0.5.0/src/github.com/opencontainers/umoci/oci/layer/xattr_linux_test.go:216
                	Error:      	Not equal: 
                	            	expected: map[string]string{"user.overlay.fakexattr":"fakexattr"}
                	            	actual  : map[string]string{"security.selinux":"system_u:object_r:mock_var_lib_t:s0\x00", "user.overlay.fakexattr":"fakexattr"}
                	            	
                	            	Diff:
                	            	--- Expected
                	            	+++ Actual
                	            	@@ -1,2 +1,3 @@
                	            	-(map[string]string) (len=1) {
                	            	+(map[string]string) (len=2) {
                	            	+ (string) (len=16) "security.selinux": (string) (len=36) "system_u:object_r:mock_var_lib_t:s0\x00",
                	            	  (string) (len=22) "user.overlay.fakexattr": (string) (len=9) "fakexattr"
                	Test:       	TestUnpackGenerateRoundTrip_ComplexXattr_OverlayfsRootfs/TarEntries=user.overlay./OverlayfsRootfs-TrustedXattr
                	Messages:   	UnpackEntry("foo/"): expected to see map[string]string{"user.overlay.fakexattr":"fakexattr"} not be remapped
            xattr_linux_test.go:216: 
                	Error Trace:	/builddir/build/BUILD/umoci-0.5.0-build/umoci-0.5.0/src/github.com/opencontainers/umoci/oci/layer/xattr_linux_test.go:216
                	Error:      	Not equal: 
                	            	expected: map[string]string{"user.overlay.whiteout":"foo"}
                	            	actual  : map[string]string{"security.selinux":"system_u:object_r:mock_var_lib_t:s0\x00", "user.overlay.whiteout":"foo"}
                	            	
                	            	Diff:
                	            	--- Expected
                	            	+++ Actual
                	            	@@ -1,2 +1,3 @@
                	            	-(map[string]string) (len=1) {
                	            	+(map[string]string) (len=2) {
                	            	+ (string) (len=16) "security.selinux": (string) (len=36) "system_u:object_r:mock_var_lib_t:s0\x00",
                	            	  (string) (len=21) "user.overlay.whiteout": (string) (len=3) "foo"
                	Test:       	TestUnpackGenerateRoundTrip_ComplexXattr_OverlayfsRootfs/TarEntries=user.overlay./OverlayfsRootfs-TrustedXattr
                	Messages:   	UnpackEntry("foo/bar"): expected to see map[string]string{"user.overlay.whiteout":"foo"} not be remapped
            xattr_linux_test.go:216: 
                	Error Trace:	/builddir/build/BUILD/umoci-0.5.0-build/umoci-0.5.0/src/github.com/opencontainers/umoci/oci/layer/xattr_linux_test.go:216
                	Error:      	Not equal: 
                	            	expected: map[string]string{"user.overlay.opaque":"y"}
                	            	actual  : map[string]string{"security.selinux":"system_u:object_r:mock_var_lib_t:s0\x00", "user.overlay.opaque":"y"}
                	            	
                	            	Diff:
                	            	--- Expected
                	            	+++ Actual
                	            	@@ -1,2 +1,3 @@
                	            	-(map[string]string) (len=1) {
                	            	+(map[string]string) (len=2) {
                	            	+ (string) (len=16) "security.selinux": (string) (len=36) "system_u:object_r:mock_var_lib_t:s0\x00",
                	            	  (string) (len=19) "user.overlay.opaque": (string) (len=1) "y"
                	Test:       	TestUnpackGenerateRoundTrip_ComplexXattr_OverlayfsRootfs/TarEntries=user.overlay./OverlayfsRootfs-TrustedXattr
                	Messages:   	UnpackEntry("foo/baz/"): expected to see map[string]string{"user.overlay.opaque":"y"} not be remapped
            xattr_linux_test.go:216: 
                	Error Trace:	/builddir/build/BUILD/umoci-0.5.0-build/umoci-0.5.0/src/github.com/opencontainers/umoci/oci/layer/xattr_linux_test.go:216
                	Error:      	Not equal: 
                	            	expected: map[string]string{"user.overlay.overlay.opaque":"x", "user.overlay.overlay.overlay.overlay.overlay.overlay.dummy":"dummy xattr", "user.overlay.overlay.overlay.whiteout":"foobar"}
                	            	actual  : map[string]string{"security.selinux":"system_u:object_r:mock_var_lib_t:s0\x00", "user.overlay.overlay.opaque":"x", "user.overlay.overlay.overlay.overlay.overlay.overlay.dummy":"dummy xattr", "user.overlay.overlay.overlay.whiteout":"foobar"}
                	            	
                	            	Diff:
                	            	--- Expected
                	            	+++ Actual
                	            	@@ -1,2 +1,3 @@
                	            	-(map[string]string) (len=3) {
                	            	+(map[string]string) (len=4) {
                	            	+ (string) (len=16) "security.selinux": (string) (len=36) "system_u:object_r:mock_var_lib_t:s0\x00",
                	            	  (string) (len=27) "user.overlay.overlay.opaque": (string) (len=1) "x",
                	Test:       	TestUnpackGenerateRoundTrip_ComplexXattr_OverlayfsRootfs/TarEntries=user.overlay./OverlayfsRootfs-TrustedXattr
                	Messages:   	UnpackEntry("foo/extra-nesting/"): expected to see map[string]string{"user.overlay.overlay.opaque":"x", "user.overlay.overlay.overlay.overlay.overlay.overlay.dummy":"dummy xattr", "user.overlay.overlay.overlay.whiteout":"foobar"} not be remapped
            xattr_linux_test.go:216: 
                	Error Trace:	/builddir/build/BUILD/umoci-0.5.0-build/umoci-0.5.0/src/github.com/opencontainers/umoci/oci/layer/xattr_linux_test.go:216
                	Error:      	Not equal: 
                	            	expected: map[string]string{"user.overlay.overlay.overlay.overlay.overlay.dummy123":"dummy xattr 123"}
                	            	actual  : map[string]string{"security.selinux":"system_u:object_r:mock_var_lib_t:s0\x00", "user.overlay.overlay.overlay.overlay.overlay.dummy123":"dummy xattr 123"}
                	            	
                	            	Diff:
                	            	--- Expected
                	            	+++ Actual
                	            	@@ -1,2 +1,3 @@
                	            	-(map[string]string) (len=1) {
                	            	+(map[string]string) (len=2) {
                	            	+ (string) (len=16) "security.selinux": (string) (len=36) "system_u:object_r:mock_var_lib_t:s0\x00",
                	            	  (string) (len=53) "user.overlay.overlay.overlay.overlay.overlay.dummy123": (string) (len=15) "dummy xattr 123"
                	Test:       	TestUnpackGenerateRoundTrip_ComplexXattr_OverlayfsRootfs/TarEntries=user.overlay./OverlayfsRootfs-TrustedXattr
                	Messages:   	UnpackEntry("foo/extra-nesting/reg"): expected to see map[string]string{"user.overlay.overlay.overlay.overlay.overlay.dummy123":"dummy xattr 123"} not be remapped
        --- FAIL: TestUnpackGenerateRoundTrip_ComplexXattr_OverlayfsRootfs/TarEntries=user.overlay./OverlayfsRootfs-UserXattr (0.01s)
            xattr_linux_test.go:206: 
                	Error Trace:	/builddir/build/BUILD/umoci-0.5.0-build/umoci-0.5.0/src/github.com/opencontainers/umoci/oci/layer/xattr_linux_test.go:206
                	Error:      	Not equal: 
                	            	expected: map[string]string{"user.dummy.xattr":"foobar", "user.overlay.overlay.opaque":"x"}
                	            	actual  : map[string]string{"security.selinux":"system_u:object_r:mock_var_lib_t:s0\x00", "user.dummy.xattr":"foobar", "user.overlay.overlay.opaque":"x"}
                	            	
                	            	Diff:
                	            	--- Expected
                	            	+++ Actual
                	            	@@ -1,2 +1,3 @@
                	            	-(map[string]string) (len=2) {
                	            	+(map[string]string) (len=3) {
                	            	+ (string) (len=16) "security.selinux": (string) (len=36) "system_u:object_r:mock_var_lib_t:s0\x00",
                	            	  (string) (len=16) "user.dummy.xattr": (string) (len=6) "foobar",
                	Test:       	TestUnpackGenerateRoundTrip_ComplexXattr_OverlayfsRootfs/TarEntries=user.overlay./OverlayfsRootfs-UserXattr
                	Messages:   	UnpackEntry("."): expected to see map[string]string{"user.dummy.xattr":"foobar", "user.overlay.opaque":"x"} remapped properly
            xattr_linux_test.go:206: 
                	Error Trace:	/builddir/build/BUILD/umoci-0.5.0-build/umoci-0.5.0/src/github.com/opencontainers/umoci/oci/layer/xattr_linux_test.go:206
                	Error:      	Not equal: 
                	            	expected: map[string]string{"user.overlay.overlay.fakexattr":"fakexattr"}
                	            	actual  : map[string]string{"security.selinux":"system_u:object_r:mock_var_lib_t:s0\x00", "user.overlay.overlay.fakexattr":"fakexattr"}
                	            	
                	            	Diff:
                	            	--- Expected
                	            	+++ Actual
                	            	@@ -1,2 +1,3 @@
                	            	-(map[string]string) (len=1) {
                	            	+(map[string]string) (len=2) {
                	            	+ (string) (len=16) "security.selinux": (string) (len=36) "system_u:object_r:mock_var_lib_t:s0\x00",
                	            	  (string) (len=30) "user.overlay.overlay.fakexattr": (string) (len=9) "fakexattr"
                	Test:       	TestUnpackGenerateRoundTrip_ComplexXattr_OverlayfsRootfs/TarEntries=user.overlay./OverlayfsRootfs-UserXattr
                	Messages:   	UnpackEntry("foo/"): expected to see map[string]string{"user.overlay.fakexattr":"fakexattr"} remapped properly
            xattr_linux_test.go:206: 
                	Error Trace:	/builddir/build/BUILD/umoci-0.5.0-build/umoci-0.5.0/src/github.com/opencontainers/umoci/oci/layer/xattr_linux_test.go:206
                	Error:      	Not equal: 
                	            	expected: map[string]string{"user.overlay.overlay.whiteout":"foo"}
                	            	actual  : map[string]string{"security.selinux":"system_u:object_r:mock_var_lib_t:s0\x00", "user.overlay.overlay.whiteout":"foo"}
                	            	
                	            	Diff:
                	            	--- Expected
                	            	+++ Actual
                	            	@@ -1,2 +1,3 @@
                	            	-(map[string]string) (len=1) {
                	            	+(map[string]string) (len=2) {
                	            	+ (string) (len=16) "security.selinux": (string) (len=36) "system_u:object_r:mock_var_lib_t:s0\x00",
                	            	  (string) (len=29) "user.overlay.overlay.whiteout": (string) (len=3) "foo"
                	Test:       	TestUnpackGenerateRoundTrip_ComplexXattr_OverlayfsRootfs/TarEntries=user.overlay./OverlayfsRootfs-UserXattr
                	Messages:   	UnpackEntry("foo/bar"): expected to see map[string]string{"user.overlay.whiteout":"foo"} remapped properly
            xattr_linux_test.go:206: 
                	Error Trace:	/builddir/build/BUILD/umoci-0.5.0-build/umoci-0.5.0/src/github.com/opencontainers/umoci/oci/layer/xattr_linux_test.go:206
                	Error:      	Not equal: 
                	            	expected: map[string]string{"user.overlay.overlay.opaque":"y"}
                	            	actual  : map[string]string{"security.selinux":"system_u:object_r:mock_var_lib_t:s0\x00", "user.overlay.overlay.opaque":"y"}
                	            	
                	            	Diff:
                	            	--- Expected
                	            	+++ Actual
                	            	@@ -1,2 +1,3 @@
                	            	-(map[string]string) (len=1) {
                	            	+(map[string]string) (len=2) {
                	            	+ (string) (len=16) "security.selinux": (string) (len=36) "system_u:object_r:mock_var_lib_t:s0\x00",
                	            	  (string) (len=27) "user.overlay.overlay.opaque": (string) (len=1) "y"
                	Test:       	TestUnpackGenerateRoundTrip_ComplexXattr_OverlayfsRootfs/TarEntries=user.overlay./OverlayfsRootfs-UserXattr
                	Messages:   	UnpackEntry("foo/baz/"): expected to see map[string]string{"user.overlay.opaque":"y"} remapped properly
            xattr_linux_test.go:206: 
                	Error Trace:	/builddir/build/BUILD/umoci-0.5.0-build/umoci-0.5.0/src/github.com/opencontainers/umoci/oci/layer/xattr_linux_test.go:206
                	Error:      	Not equal: 
                	            	expected: map[string]string{"user.overlay.overlay.overlay.opaque":"x", "user.overlay.overlay.overlay.overlay.overlay.overlay.overlay.dummy":"dummy xattr", "user.overlay.overlay.overlay.overlay.whiteout":"foobar"}
                	            	actual  : map[string]string{"security.selinux":"system_u:object_r:mock_var_lib_t:s0\x00", "user.overlay.overlay.overlay.opaque":"x", "user.overlay.overlay.overlay.overlay.overlay.overlay.overlay.dummy":"dummy xattr", "user.overlay.overlay.overlay.overlay.whiteout":"foobar"}
                	            	
                	            	Diff:
                	            	--- Expected
                	            	+++ Actual
                	            	@@ -1,2 +1,3 @@
                	            	-(map[string]string) (len=3) {
                	            	+(map[string]string) (len=4) {
                	            	+ (string) (len=16) "security.selinux": (string) (len=36) "system_u:object_r:mock_var_lib_t:s0\x00",
                	            	  (string) (len=35) "user.overlay.overlay.overlay.opaque": (string) (len=1) "x",
                	Test:       	TestUnpackGenerateRoundTrip_ComplexXattr_OverlayfsRootfs/TarEntries=user.overlay./OverlayfsRootfs-UserXattr
                	Messages:   	UnpackEntry("foo/extra-nesting/"): expected to see map[string]string{"user.overlay.overlay.opaque":"x", "user.overlay.overlay.overlay.overlay.overlay.overlay.dummy":"dummy xattr", "user.overlay.overlay.overlay.whiteout":"foobar"} remapped properly
            xattr_linux_test.go:206: 
                	Error Trace:	/builddir/build/BUILD/umoci-0.5.0-build/umoci-0.5.0/src/github.com/opencontainers/umoci/oci/layer/xattr_linux_test.go:206
                	Error:      	Not equal: 
                	            	expected: map[string]string{"user.overlay.overlay.overlay.overlay.overlay.overlay.dummy123":"dummy xattr 123"}
                	            	actual  : map[string]string{"security.selinux":"system_u:object_r:mock_var_lib_t:s0\x00", "user.overlay.overlay.overlay.overlay.overlay.overlay.dummy123":"dummy xattr 123"}
                	            	
                	            	Diff:
                	            	--- Expected
                	            	+++ Actual
                	            	@@ -1,2 +1,3 @@
                	            	-(map[string]string) (len=1) {
                	            	+(map[string]string) (len=2) {
                	            	+ (string) (len=16) "security.selinux": (string) (len=36) "system_u:object_r:mock_var_lib_t:s0\x00",
                	            	  (string) (len=61) "user.overlay.overlay.overlay.overlay.overlay.overlay.dummy123": (string) (len=15) "dummy xattr 123"
                	Test:       	TestUnpackGenerateRoundTrip_ComplexXattr_OverlayfsRootfs/TarEntries=user.overlay./OverlayfsRootfs-UserXattr
                	Messages:   	UnpackEntry("foo/extra-nesting/reg"): expected to see map[string]string{"user.overlay.overlay.overlay.overlay.overlay.dummy123":"dummy xattr 123"} remapped properly
--- FAIL: TestUnpackGenerateRoundTrip_MockedSELinux (0.01s)
    --- FAIL: TestUnpackGenerateRoundTrip_MockedSELinux/DirRootfs (0.00s)
        xattr_linux_test.go:351: 
            	Error Trace:	/builddir/build/BUILD/umoci-0.5.0-build/umoci-0.5.0/src/github.com/opencontainers/umoci/oci/layer/xattr_linux_test.go:351
            	Error:      	Not equal: 
            	            	expected: map[string]string{"user.UMOCI.fake_selinux":"rootdir", "user.dummy.xattr":"foobar"}
            	            	actual  : map[string]string{"security.selinux":"system_u:object_r:mock_var_lib_t:s0\x00", "user.UMOCI.fake_selinux":"rootdir", "user.dummy.xattr":"foobar"}
            	            	
            	            	Diff:
            	            	--- Expected
            	            	+++ Actual
            	            	@@ -1,2 +1,3 @@
            	            	-(map[string]string) (len=2) {
            	            	+(map[string]string) (len=3) {
            	            	+ (string) (len=16) "security.selinux": (string) (len=36) "system_u:object_r:mock_var_lib_t:s0\x00",
            	            	  (string) (len=23) "user.UMOCI.fake_selinux": (string) (len=7) "rootdir",
            	Test:       	TestUnpackGenerateRoundTrip_MockedSELinux/DirRootfs
            	Messages:   	UnpackEntry("."): expected to only see specific subset of applied xattrs
        xattr_linux_test.go:351: 
            	Error Trace:	/builddir/build/BUILD/umoci-0.5.0-build/umoci-0.5.0/src/github.com/opencontainers/umoci/oci/layer/xattr_linux_test.go:351
            	Error:      	Not equal: 
            	            	expected: map[string]string{"user.UMOCI.fake_selinux":"foodir", "user.dummy.xattr":"barbaz"}
            	            	actual  : map[string]string{"security.selinux":"system_u:object_r:mock_var_lib_t:s0\x00", "user.UMOCI.fake_selinux":"foodir", "user.dummy.xattr":"barbaz"}
            	            	
            	            	Diff:
            	            	--- Expected
            	            	+++ Actual
            	            	@@ -1,2 +1,3 @@
            	            	-(map[string]string) (len=2) {
            	            	+(map[string]string) (len=3) {
            	            	+ (string) (len=16) "security.selinux": (string) (len=36) "system_u:object_r:mock_var_lib_t:s0\x00",
            	            	  (string) (len=23) "user.UMOCI.fake_selinux": (string) (len=6) "foodir",
            	Test:       	TestUnpackGenerateRoundTrip_MockedSELinux/DirRootfs
            	Messages:   	UnpackEntry("foo/"): expected to only see specific subset of applied xattrs
        xattr_linux_test.go:351: 
            	Error Trace:	/builddir/build/BUILD/umoci-0.5.0-build/umoci-0.5.0/src/github.com/opencontainers/umoci/oci/layer/xattr_linux_test.go:351
            	Error:      	Not equal: 
            	            	expected: map[string]string{"user.UMOCI.fake_selinux":"foobarfile"}
            	            	actual  : map[string]string{"security.selinux":"system_u:object_r:mock_var_lib_t:s0\x00", "user.UMOCI.fake_selinux":"foobarfile"}
            	            	
            	            	Diff:
            	            	--- Expected
            	            	+++ Actual
            	            	@@ -1,2 +1,3 @@
            	            	-(map[string]string) (len=1) {
            	            	+(map[string]string) (len=2) {
            	            	+ (string) (len=16) "security.selinux": (string) (len=36) "system_u:object_r:mock_var_lib_t:s0\x00",
            	            	  (string) (len=23) "user.UMOCI.fake_selinux": (string) (len=10) "foobarfile"
            	Test:       	TestUnpackGenerateRoundTrip_MockedSELinux/DirRootfs
            	Messages:   	UnpackEntry("foo/bar"): expected to only see specific subset of applied xattrs
    --- FAIL: TestUnpackGenerateRoundTrip_MockedSELinux/OverlayfsRootfs (0.00s)
        xattr_linux_test.go:351: 
            	Error Trace:	/builddir/build/BUILD/umoci-0.5.0-build/umoci-0.5.0/src/github.com/opencontainers/umoci/oci/layer/xattr_linux_test.go:351
            	Error:      	Not equal: 
            	            	expected: map[string]string{"user.UMOCI.fake_selinux":"rootdir", "user.dummy.xattr":"foobar"}
            	            	actual  : map[string]string{"security.selinux":"system_u:object_r:mock_var_lib_t:s0\x00", "user.UMOCI.fake_selinux":"rootdir", "user.dummy.xattr":"foobar"}
            	            	
            	            	Diff:
            	            	--- Expected
            	            	+++ Actual
            	            	@@ -1,2 +1,3 @@
            	            	-(map[string]string) (len=2) {
            	            	+(map[string]string) (len=3) {
            	            	+ (string) (len=16) "security.selinux": (string) (len=36) "system_u:object_r:mock_var_lib_t:s0\x00",
            	            	  (string) (len=23) "user.UMOCI.fake_selinux": (string) (len=7) "rootdir",
            	Test:       	TestUnpackGenerateRoundTrip_MockedSELinux/OverlayfsRootfs
            	Messages:   	UnpackEntry("."): expected to only see specific subset of applied xattrs
        xattr_linux_test.go:351: 
            	Error Trace:	/builddir/build/BUILD/umoci-0.5.0-build/umoci-0.5.0/src/github.com/opencontainers/umoci/oci/layer/xattr_linux_test.go:351
            	Error:      	Not equal: 
            	            	expected: map[string]string{"user.UMOCI.fake_selinux":"foodir", "user.dummy.xattr":"barbaz"}
            	            	actual  : map[string]string{"security.selinux":"system_u:object_r:mock_var_lib_t:s0\x00", "user.UMOCI.fake_selinux":"foodir", "user.dummy.xattr":"barbaz"}
            	            	
            	            	Diff:
            	            	--- Expected
            	            	+++ Actual
            	            	@@ -1,2 +1,3 @@
            	            	-(map[string]string) (len=2) {
            	            	+(map[string]string) (len=3) {
            	            	+ (string) (len=16) "security.selinux": (string) (len=36) "system_u:object_r:mock_var_lib_t:s0\x00",
            	            	  (string) (len=23) "user.UMOCI.fake_selinux": (string) (len=6) "foodir",
            	Test:       	TestUnpackGenerateRoundTrip_MockedSELinux/OverlayfsRootfs
            	Messages:   	UnpackEntry("foo/"): expected to only see specific subset of applied xattrs
        xattr_linux_test.go:351: 
            	Error Trace:	/builddir/build/BUILD/umoci-0.5.0-build/umoci-0.5.0/src/github.com/opencontainers/umoci/oci/layer/xattr_linux_test.go:351
            	Error:      	Not equal: 
            	            	expected: map[string]string{"user.UMOCI.fake_selinux":"foobarfile"}
            	            	actual  : map[string]string{"security.selinux":"system_u:object_r:mock_var_lib_t:s0\x00", "user.UMOCI.fake_selinux":"foobarfile"}
            	            	
            	            	Diff:
            	            	--- Expected
            	            	+++ Actual
            	            	@@ -1,2 +1,3 @@
            	            	-(map[string]string) (len=1) {
            	            	+(map[string]string) (len=2) {
            	            	+ (string) (len=16) "security.selinux": (string) (len=36) "system_u:object_r:mock_var_lib_t:s0\x00",
            	            	  (string) (len=23) "user.UMOCI.fake_selinux": (string) (len=10) "foobarfile"
            	Test:       	TestUnpackGenerateRoundTrip_MockedSELinux/OverlayfsRootfs
            	Messages:   	UnpackEntry("foo/bar"): expected to only see specific subset of applied xattrs
FAIL
FAIL	github.com/opencontainers/umoci/oci/layer	1.159s
?   	github.com/opencontainers/umoci/pkg/fseval	[no test files]
ok  	github.com/opencontainers/umoci/pkg/funchelpers	0.007s
ok  	github.com/opencontainers/umoci/pkg/hardening	0.027s
ok  	github.com/opencontainers/umoci/pkg/idtools	0.009s
?   	github.com/opencontainers/umoci/pkg/iohelpers	[no test files]
ok  	github.com/opencontainers/umoci/pkg/mtreefilter	0.018s
ok  	github.com/opencontainers/umoci/pkg/pathtrie	0.008s
--- FAIL: TestClearxattrFilter (0.00s)
    xattr_unix_test.go:72: 
        	Error Trace:	/builddir/build/BUILD/umoci-0.5.0-build/umoci-0.5.0/src/github.com/opencontainers/umoci/pkg/system/xattr_unix_test.go:72
        	Error:      	elements differ
        	            	
        	            	extra elements in list B:
        	            	([]interface {}) (len=1) {
        	            	 (string) (len=16) "security.selinux"
        	            	}
        	            	
        	            	
        	            	listA:
        	            	([]string) (len=4) {
        	            	 (string) (len=13) "user.allowed1",
        	            	 (string) (len=13) "user.allowed2",
        	            	 (string) (len=15) "user.forbidden1",
        	            	 (string) (len=23) "user.forbidden1.allowed"
        	            	}
        	            	
        	            	
        	            	listB:
        	            	([]string) (len=5) {
        	            	 (string) (len=23) "user.forbidden1.allowed",
        	            	 (string) (len=13) "user.allowed2",
        	            	 (string) (len=13) "user.allowed1",
        	            	 (string) (len=15) "user.forbidden1",
        	            	 (string) (len=16) "security.selinux"
        	            	}
        	Test:       	TestClearxattrFilter
        	Messages:   	all xattrs should be present after setting
    xattr_unix_test.go:85: 
        	Error Trace:	/builddir/build/BUILD/umoci-0.5.0-build/umoci-0.5.0/src/github.com/opencontainers/umoci/pkg/system/xattr_unix_test.go:85
        	Error:      	elements differ
        	            	
        	            	extra elements in list B:
        	            	([]interface {}) (len=1) {
        	            	 (string) (len=16) "security.selinux"
        	            	}
        	            	
        	            	
        	            	listA:
        	            	([]string) (len=1) {
        	            	 (string) (len=15) "user.forbidden1"
        	            	}
        	            	
        	            	
        	            	listB:
        	            	([]string) (len=2) {
        	            	 (string) (len=15) "user.forbidden1",
        	            	 (string) (len=16) "security.selinux"
        	            	}
        	Test:       	TestClearxattrFilter
        	Messages:   	only explicitly forbidden xattrs should be allowed to remain after clearing
FAIL
FAIL	github.com/opencontainers/umoci/pkg/system	0.016s
?   	github.com/opencontainers/umoci/pkg/testutils	[no test files]
ok  	github.com/opencontainers/umoci/pkg/unpriv	0.043s
FAIL
```

--- a/internal/system/xattr_unix_test.go
+++ b/internal/system/xattr_unix_test.go
@@ -28,6 +28,16 @@ import (
 	"golang.org/x/sys/unix"
 )
 
+// specialXattrs come from the OCI xattr remapping logic in oci/layer/xattr.go.
+// These are xattrs that will be auto-set by the system and we should ignore
+// their existence when doing xattr-related operations (for our tests, this
+// especially means to take care of them when checking sets of xattrs that
+// exist on the filesystem).
+var specialXattrs = map[string]struct{}{
+	"security.selinux": {},
+	"system.nfs4_acl":  {},
+}
+
 func TestClearxattrFilter(t *testing.T) {
 	dir := t.TempDir()
 
@@ -37,6 +47,12 @@ func TestClearxattrFilter(t *testing.T) {
 
 	path := file.Name()
 	defer os.RemoveAll(path) //nolint:errcheck
+
+	autosetXattrs, err := Llistxattr(path)
+	require.NoErrorf(t, err, "llistxattr %q", path)
+	for _, xattr := range autosetXattrs {
+		require.Contains(t, specialXattrs, xattr, "auto-set xattrs must be part of special list")
+	}
 
 	xattrs := []struct {
 		name, value string
@@ -65,11 +81,16 @@ func TestClearxattrFilter(t *testing.T) {
 		}
 		require.NoErrorf(t, err, "lsetxattr %q=%q on %q", xattr.name, xattr.value, path)
 	}
+	// If we are running on an SELinux-enabled system, all new files get a
+	// security.selinux xattr that gets auto-set and cannot be removed so we
+	// need to include it in the expected set.
+	expectAllXattrNames := append(setXattrNames, autosetXattrs...)
+	expectRemainingXattrNames := append(forbiddenXattrNames, autosetXattrs...)
 
 	// Check they're all present.
 	allXattrList, err := Llistxattr(path)
 	require.NoErrorf(t, err, "llistxattr %q", path)
-	assert.ElementsMatch(t, setXattrNames, allXattrList, "all xattrs should be present after setting")
+	assert.ElementsMatch(t, expectAllXattrNames, allXattrList, "all xattrs should be present after setting")
 
 	// Now clear them.
 	err = Lclearxattrs(path, func(xattrName string) bool {
@@ -79,9 +100,9 @@ func TestClearxattrFilter(t *testing.T) {
 	require.NoErrorf(t, err, "lclearxattrs %q (forbidden=%v)", path, forbiddenXattrs)
 
 	// Check that only the forbidden ones remain.
-	forbiddenXattrList, err := Llistxattr(path)
+	remainingXattrList, err := Llistxattr(path)
 	require.NoErrorf(t, err, "llistxattr %q", path)
-	assert.NotElementsMatch(t, setXattrNames, forbiddenXattrList, "there should be a different set of xattrs after clearing")
-	assert.ElementsMatch(t, forbiddenXattrNames, forbiddenXattrList, "only explicitly forbidden xattrs should be allowed to remain after clearing")
-	assert.NotEmpty(t, forbiddenXattrList, "there should be some remaining xattrs after clearing")
+	assert.NotElementsMatch(t, expectAllXattrNames, remainingXattrList, "there should be a different set of xattrs after clearing")
+	assert.ElementsMatch(t, expectRemainingXattrNames, remainingXattrList, "only explicitly forbidden xattrs should be allowed to remain after clearing")
+	assert.NotEmpty(t, remainingXattrList, "there should be some remaining xattrs after clearing")
 }

--- a/oci/layer/xattr_linux_test.go
+++ b/oci/layer/xattr_linux_test.go
@@ -37,6 +37,11 @@ import (
 	"github.com/opencontainers/umoci/pkg/fseval"
 )
 
+func isForbiddenXattr(xattrName string) bool {
+	found, ok := getXattrFilter(xattrName)
+	return ok && found == forbiddenXattrFilter{}
+}
+
 func getAllXattrs(t *testing.T, path string) map[string]string {
 	names, err := system.Llistxattr(path)
 	require.NoErrorf(t, err, "fetch all xattrs for %q", path)
@@ -201,6 +206,14 @@ func testUnpackGenerateRoundTrip_ComplexXattr_OverlayfsRootfs(t *testing.T, tarX
 				fullPath := filepath.Join(dir, path)
 
 				xattrs := getAllXattrs(t, fullPath)
+				// Strip out any forbidden xattrs from the returned set.
+				// TODO: This is a little ugly -- ideally we would create
+				// copies of the necessary maps with special xattrs added.
+				for name := range xattrs {
+					if isForbiddenXattr(name) {
+						delete(xattrs, name)
+					}
+				}
 
 				if test.expectRemap {
 					assert.Equalf(t, de.remapXattrs, xattrs, "UnpackEntry(%q): expected to see %#v remapped properly", path, de.xattrs)
@@ -212,7 +225,6 @@ func testUnpackGenerateRoundTrip_ComplexXattr_OverlayfsRootfs(t *testing.T, tarX
 					require.NoErrorf(t, err, "isOverlayWhiteout(%q)", path)
 					assert.Falsef(t, isWo, "isOverlayWhiteout(%q): regular entries with overlayfs xattrs should not end up being unpacked with overlayfs whiteout xattrs", path)
 				} else {
-					xattrs := getAllXattrs(t, fullPath)
 					assert.Equalf(t, de.xattrs, xattrs, "UnpackEntry(%q): expected to see %#v not be remapped", path, de.xattrs)
 					assert.NotEqualf(t, de.remapXattrs, xattrs, "UnpackEntry(%q): expected to see %#v not be remapped", path, de.xattrs)
 				}
@@ -347,6 +359,14 @@ func TestUnpackGenerateRoundTrip_MockedSELinux(t *testing.T) {
 				// forbidden xattr after all the extractions.
 				if value, ok := de.autoXattrs[forbiddenTestXattr]; ok {
 					wantXattrs[forbiddenTestXattr] = value
+				}
+				// If we are running on an SELinux system (or even more
+				// unlikely, on top of nfsv4) we need to also allow any masked
+				// xattrs auto-applied by the system.
+				for gotXattr, value := range xattrs {
+					if isForbiddenXattr(gotXattr) {
+						wantXattrs[gotXattr] = value
+					}
 				}
 				assert.Equalf(t, wantXattrs, xattrs, "UnpackEntry(%q): expected to only see specific subset of applied xattrs", path)
 			}

